### PR TITLE
Recommend to baseline after (but not before) ICA; and warn otherwise; warn if data was not high-passed when fitting ICA

### LINKED ISSUE #9021: RFC: Re-applying baseline correction after ICA
Skimming over some Study Template analysis results with @agramfort, we found that some of our data appeared to be not baseline-corrected – some traces clearly didn't have a zero-mean during the baseline period.

So I started digging and figured that we had **indeed** applied baseline correction to our data; yet after cleaning the Epochs via ICA, the baseline signals of some channels had started floating away from zero.

Now this may not come as a surprise to some of you, but I have to admit it caught me off-guard. I'd never seen this behavior before with the data I'd worked with so far, and I've failed to reproduce a similarly visible effect using our `sample` data.

However, looking through the [ICA tutorial](https://mne.tools/dev/auto_tutorials/preprocessing/plot_40_artifact_correction_ica.html#fitting-and-plotting-the-ica-solution), where we clean Raw data, one can see the essence of this effect too – albeit with Raw and not Epochs (lower panel):

I'd like to propose and discuss the following changes and enhancements to deal with this behavior:

- Clearly state in the docs that baseline correction should be re-applied to Epochs and Evoked data after applying ICA
- `viz.plot_ica_overlay()` should automatically re-apply baseline correction to the cleaned signal if an Evoked was passed
    - for this, we'd need to add an `Evoked.baseline` attribute, which gets inherited from the parent `Epochs.baseline`. This has been on my agenda for a while anyway in order to improve provenance, so I'd might take a shot at this anyway.
- We could even go so far as to auto-reapply baseline correction to Epochs and Evoked data as part of `ICA.apply()`, so users wouldn't have to do this themselves. This behavior should be toggle-able using a new kwarg.

### PR DESCRIPTION
**Pull request by hoechenberger**
Closes #9021

### CODE REVIEW

**Review by larsoner — APPROVED** [review]

**Comment by larsoner** [comment]
... whoops -- too soon, needs tests.

File: mne/preprocessing/ica.py [hunk 1]
```diff
@@ -1509,16 +1522,25 @@ def apply(self, inst, include=None, exclude=None, n_pca_components=None,
...
         inst : instance of Raw, Epochs or Evoked
-            The data to be processed. The instance is modified inplace.
+            The data to be processed (i.e., cleaned). It will be modified
+            inplace.
+
+            .. note:: In case of `~mne.Epochs` and `~mne.Evoked`, it is
+                      recommended to pass data that has **not** been
+                      baseline-corrected. A warning will be emitted otherwise
+                      in the case of `~mne.Evoked`. If you require baseline
+                      correction, do it **after** cleaning the data.
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/preprocessing/ica.py:1540]: oh! :)

[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/preprocessing/ica.py:1540]: I **thought** I remembered differently … but would have to check again!

[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/preprocessing/ica.py:1540]: Ok so I did a test with real data (ERP CORE).

- ICA fitted on highpass filtered epochs w/o baseline correction
- then after exclusion of a few ICs, `ICA.apply()` on baseline-corrected epochs

result:
Screen Shot 2021-03-12 at 17 39 16

This clearly needs to be baseline-corrected again.

So maybe the note and emitted warning should be rephrased to only indicate that one needs to re-apply baseline correction? And the part about that we don't recommend running `ICA.apply()` on baselined data should be removed?

[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/preprocessing/ica.py]: @agramfort I've rephrased things, can you have another look?

**Comment by hoechenberger** [comment]
This is ready for review.

We now warn:

- If the data passed to `ICA.fit()` was not high-passed (suggest high-pass)
- If the data passed to `ICA.fit()` was baselined (suggest not baselining)
- If the data passed to `ICA.apply()` was baselined (suggest re-applying baseline to remove DC shifts)

**Comment by agramfort** [comment]
@hoechenberger can you push a commit to with circle full build to make sure we don't break the doc?

**Comment by hoechenberger** [comment]
thanks @larsoner!

**Comment by larsoner** [comment]
Good thing we ran `circle full`!
```
Traceback (most recent call last):
  File "/home/circleci/project/examples/preprocessing/plot_run_ica.py", line 49, in <module>
    ica = ICA(n_components=0.95, method='fastica').fit(epochs)
  File "<decorator-gen-414>", line 24, in fit
  File "/home/circleci/project/mne/preprocessing/ica.py", line 518, in fit
    warn('The epochs you passed to ICA.fit() were baseline-corrected. '
  File "/home/circleci/project/mne/utils/_logging.py", line 358, in warn
    warnings.warn_explicit(
RuntimeWarning: The epochs you passed to ICA.fit() were baseline-corrected. However, we suggest to fit ICA only on data that has been high-pass filtered, but NOT baseline-corrected.
```

**Comment by hoechenberger** [comment]
Oh wow!!

What does circle full do, exactly? Try to build all docs, examples, and tutorials?

**Comment by agramfort** [comment]
yes it builds every example not just the ones with some modifications

File: mne/preprocessing/ica.py [hunk 2]
```diff
@@ -1529,16 +1547,27 @@ def apply(self, inst, include=None, exclude=None, n_pca_components=None,
...
+            .. note:: Applying ICA may introduce a DC shift. If you pass
+                      baseline-corrected `~mne.Epochs` or `~mne.Evoked` data,
+                      the baseline period of the cleaned data may not be of
+                      zero mean anymore. If you require baseline-corrected
+                      data, apply baseline correction again after cleaning
+                      via ICA. A warning will be emitted to remind you of this
+                      fact if you pass baseline-corrected data.
+
+            .. versionchanged:: 0.23
+               Warn if instance was baseline-corrected.
```
Comment (larsoner) [inline mne/preprocessing/ica.py:1570]: This is not really a change in the parameter itself but rather the behavior / handling of the parameter, so this is probably better in the `Notes` section.

I would also put the `.. note::` in the notes section.

Same for the above docstring changes

**Review by larsoner — COMMENTED** [review]
The one use case that I think *might* be problematic is fitting and applying ICA on the same instance of epochs that have not been lowpassed but have been baseline corrected. I think in this case everything should be okay (some DC has been accounted for, `epochs = ica.fit(epochs).apply(epochs)` should still be zero-mean during the baseline, and I think it should work okay).

In fact, can you check (add a test) that it's the case that non-high-passed but with-baseline-correction epochs has that `epochs = ica.fit(epochs).apply(epochs)` still has zero during the baseline period? I think we'll want the test either way, and we should think about if this should emit two warnings (as it does in this PR), as it maybe doesn't emit any in `main`.

One thing we could consider doing is checking to see if `epochs.baseline` was non-None during fitting (by saving it?) and just checking to see if the object being `.apply`ed has the same baseline or not...

**Comment by hoechenberger** [comment]
> One thing we could consider doing is checking to see if `epochs.baseline` was non-None during fitting (by saving it?) and just checking to see if the object being `.apply`ed has the same baseline or not...

To ensure this new `ICA` attribute survives an I/O roundtrip, we'd need to add a new FIFF constant, or am I missing something?

**Comment by larsoner** [comment]
> To ensure this new ICA attribute survives an I/O roundtrip, we'd need to add a new FIFF constant, or am I missing something?

We can could just reuse an existing one. For example `FIFF.FIFF_MNE_BASELINE_MIN` / `MAX` can be used within the ICA block, so no confusion/collision with it being used inside an epochs block within a given file (though we never really write an ICA block and an Epochs block to the same file anyway).

**Comment by hoechenberger** [comment]
Ok. Before starting work on this, I'll write the test you proposed. Just to be on the safe side ;)

**Comment by hoechenberger** [comment]
> In fact, can you check (add a test) that it's the case that non-high-passed but with-baseline-correction epochs has that `epochs = ica.fit(epochs).apply(epochs)` still has zero during the baseline period? I think we'll want the test either way, and we should think about if this should emit two warnings (as it does in this PR), as it maybe doesn't emit any in `main`.

If we remove independent components, the result will need to be baseline-corrected again:

```python
# %%
import os
import numpy as np
import mne
from mne.preprocessing import ICA

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
event_id = {'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3,
            'visual/right': 4, 'face': 5, 'button': 32}

raw = mne.io.read_raw_fif(sample_data_raw_file).pick_types(meg=False, eeg=True,
                                                           stim=True)
events = mne.find_events(raw, stim_channel='STI 014')

epochs = mne.Epochs(raw, events, tmin=-0.2, tmax=0.5,
                    event_id=event_id, preload=True, baseline=(None, 0))
epochs.pick_types(eeg=True)

ica = ICA(n_components=15, max_iter='auto', random_state=97)
ica.fit(epochs)
ica.exclude = [0, 1, 3]

epochs_cleaned = ica.apply(epochs.copy())

t0_idx = epochs_cleaned.time_as_index(0)[0]
baseline_data = epochs_cleaned.get_data()[:, :, :t0_idx+1]
max_abs_val = np.abs(baseline_data.mean(axis=-1)).max()

print(f'Max. absolute value of cleaned evoked during baseline: {max_abs_val}')


evoked = epochs.average()
fig = ica.plot_overlay(evoked)
fig.tight_layout()
```

```
Max. absolute value of cleaned evoked during baseline: 2.0290096363699685e-06
```
Screen Shot 2021-03-23 at 20 48 40

**Comment by hoechenberger** [comment]
Friendly pin @larsoner :)
It seems that baseline correcting again after artifact rejection seems to be a good idea in general. (see my comment above)

-> I think informing users to apply BL correction anew after `ICA.apply()` is okay

However, maybe we don't need to warn if they **fit** ICA on data that was **only** BL-corrected, but not low-pass filtered (currently we suggest that users low-pass the data)

Thoughts?

**Comment by larsoner** [comment]
Yep, I think I was just wrong about the baseline corrected case. So I think my only remaining suggestion is to move the new note/versionadded to notes!

**Comment by hoechenberger** [comment]
> Yep, I think I was just wrong about the baseline corrected case. So I think my only remaining suggestion is to move the new note/versionadded to notes!

What's your opinion on `ICA.fit()` on baselined Epochs that were created from unfiltered Raw?

**Comment by larsoner** [comment]
> What's your opinion on ICA.fit() on baselined Epochs that were created from unfiltered Raw?

This is what your example above is, and it seems to be problematic there, right?

**Comment by hoechenberger** [comment]
> > What's your opinion on ICA.fit() on baselined Epochs that were created from unfiltered Raw?
> 
> This is what your example above is, and it seems to be problematic there, right?

It's problematic at the reconstruction stage alright, but I'm not sure if it warrants to warn at the **fitting / decomposition** stage about "hey, your data was not high-passed". I wonder if with short time intervals like these – i.e., epochs – baseline-correction is maybe just as good as high-passing before fitting ICA… but I'm not sure?

**Comment by larsoner** [comment]
> It's problematic at the reconstruction stage alright, but I'm not sure if it warrants to warn at the fitting / decomposition stage about "hey, your data was not high-passed"

I think it's okay, because if the data are not high passed then this wackiness with the means and ICA not acting like a linear transformation is just going to be worse

**Comment by hoechenberger** [comment]
So this is the tutorial artifact:
https://26591-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/plot_40_artifact_correction_ica.html

@drammock I realized that the headers displayed in the right sidebar are not really helpful in this long tutorial: they stop at a too high level. Any suggestion on how to change this?

Somewhat OT: While I appreciate consolidation of examples and tutorials, I think this particular tutorial is now way too big … I just ran it locally and was totally overwhelmed, so many figures popping up, I didn't really know what was going on anymore. Maybe we should consider splitting up long tutorials into multiple smaller ones.

**Comment by drammock** [comment]
> @drammock I realized that the headers displayed in the right sidebar are not really helpful for navigation in this long tutorial: they stop at a too high nesting level. Any suggestion on how to change this?

I noticed that too when doing my final review of @neurogima's additions to that tutorial.  I think the right solution is to convert most/all the level-2 section headings to level 1 headings.

> I think this particular tutorial is now way too big

It is extremely big. But it has the advantage of the "ICA on epochs" section re-using objects that already exist, so I think the overall memory footprint is lower.  If we are going to split out part of it, I would suggest to split out the part that uses a different dataset (i.e., the template matching sections)

**Comment by drammock** [comment]
I would also add that:

1. ICA is an extremely important and widely-used preprocessing strategy so it probably deserves a lot of tutorial real-estate (though I don't think you were disagreeing with that @hoechenberger).  
2. I don't normally think of folks running these tutorials *as a script* but rather running them cell by cell (as a notebook) or line-by-line (as a `.py` script in their IDE, using a "send line to console" keyboard shortcut probably).  When done that way, you don't get a ton of figures all at once, just one or a few.  If running-as-a-script is something we want to tailor for, it would really change the way I approach these tutorials (i.e., they'd be more like our how-to examples: short, only generate one figure). But I don't think that's the right approach for these tutorials. Do you disagree?

**Comment by hoechenberger** [comment]
Thanks @drammock, I agree with everything you said. I will try to address the issue with the heading levels in a followup PR.

**Comment by hoechenberger** [comment]
Whoops, seems I merged too early? @larsoner @drammock are you happy with the changes? Or should I revert?

**Comment by larsoner** [comment]
I think we're good!

**Pull request merged**